### PR TITLE
Avoid dangling deletes if disk not found

### PR DESCRIPTION
**Pull request by poornas**
If any of the drives returns disk not found error, ignore dangling delete logic

## Motivation and Context
Avoid stale object logic if all drives are not online

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :x: [more...](#73e321eb-3f4e-4056-8e68-26e5b0cde551) |
<a name="73e321eb-3f4e-4056-8e68-26e5b0cde551"></a>
### 16401-2d26852/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.8:31483
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 3c4f5502dc5c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 16 seconds
(8/14) Running minio-go tests ... done in 48 seconds
(9/14) Running minio-java tests ... FAILED in 27 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 361,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"17396FAAF2CE8D3E\",\"HostId\":\"dbf30a18-1963-4350-b2df-debab6da1fb6\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 49 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 40 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
Looks like we re-reproduced this problem that was supposed to be fixed.

I am not sure when this was removed. Without this change we have disastrous consequences upon quorum failures during GET

**Comment by harshavardhana** [comment]
@vadmeste @klauspost ^^ PTAL

**Comment by vadmeste** [comment]
@poornas I wasn't able to reproduce for now, can you tell how to reproduce this ?

But I am okay anyway to add extra defensive code here.

**Comment by harshavardhana** [comment]
Took the high road - dangling deletes are no longer needed #16404 for most part 

```
commit cc1d8f005737269c3f3b9e7808334f5c8e974193
Author: Klaus Post <klauspost@gmail.com>
Date:   Mon Nov 28 19:20:55 2022 +0100

    Check for abandoned data when healing (#16122)
```

This code safely handles most situations.

**Comment by klauspost** [comment]
@poornas 

I think it is actually working as-is, since it sees `errDiskNotFound`, `errUnformattedDisk` as valid data, and therefore assumes it is good.

This allows for dangling deletes to be performed when there isn't enough data or parts, even if the missing disks are brought back.

Looking at the code, there isn't any correlation done between `errs[idx]` and `dataErrs[idx]`, meaning that one broken disk could potentially count as double. However it seems like all call paths have `dataErrs == nil`, so maybe that should just be cleaned up.

**Comment by vadmeste** [comment]
> totalErrs := notFoundMetaErrs + corruptedMetaErrs + notFoundPartsErrs + corruptedPartsErrs

can we have corrupted meta when there is a network issue during remote reading of xl.meta ? I can try it if you think it is possible

**Comment by klauspost** [comment]
Not really. It would be another error then.

**Comment by poornas** [comment]
> @poornas I wasn't able to reproduce for now, can you tell how to reproduce this ?
> 
> But I am okay anyway to add extra defensive code here.

Run `confess --access-key minio --secret-key minio123 --bucket "bucket" -o /tmp/dir http://localhost:900\{1...2\}` where bucket has replication enabled. Repeatedly doing `mc admin service restart <alias>` shows `The specified key does not exist`  on Get|Stat call around restart - in this set up the first Get/Head is successful on one of the nodes, but is deleted as stale before the second GET/HEAD

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/erasure-healing.go [hunk 1]
```diff
@@ -963,9 +966,13 @@ func isObjectDangling(metaArr []FileInfo, errs []error, dataErrs []error) (valid
 		}
 	}
 
-	notFoundMetaErrs, corruptedMetaErrs := danglingErrsCount(errs)
-	notFoundPartsErrs, corruptedPartsErrs := danglingErrsCount(ndataErrs)
+	notFoundMetaErrs, corruptedMetaErrs, driveNotFoundMetaErrs := danglingErrsCount(errs)
+	notFoundPartsErrs, corruptedPartsErrs, driveNotFoundPartsErrs := danglingErrsCount(ndataErrs)
 
+	if driveNotFoundMetaErrs > 0 || driveNotFoundPartsErrs > 0 {
+		fmt.Println("skipped dangling .....because", errs, "|", ndataErrs)
```
Comment (harshavardhana) [inline cmd/erasure-healing.go]: Do we need this log here ?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**